tests/integration/tests/test_util/harness/lxd.py

```diff
@@ -29,53 +29,60 @@ def __init__(self):
         self._next_id = 0
 
         self.profile = config.LXD_PROFILE_NAME
+        self.dualstack_profile = None
         self.sideload_images_dir = config.LXD_SIDELOAD_IMAGES_DIR
         self.image = config.LXD_IMAGE
         self.instances = set()
 
-        LOG.debug("Checking for LXD profile %s", self.profile)
-        try:
-            run(["lxc", "profile", "show", self.profile])
-        except subprocess.CalledProcessError:
-            try:
-                LOG.debug("Creating LXD profile %s", self.profile)
-                run(["lxc", "profile", "create", self.profile])
-
-            except subprocess.CalledProcessError as e:
-                raise HarnessError(
-                    f"Failed to create LXD profile {self.profile}"
-                ) from e
-
-        try:
-            LOG.debug("Configuring LXD profile %s", self.profile)
-            run(
-                ["lxc", "profile", "edit", self.profile],
-                input=config.LXD_PROFILE.encode(),
-            )
-        except subprocess.CalledProcessError as e:
-            raise HarnessError(f"Failed to configure LXD profile {self.profile}") from e
+        configure_lxd_profile(self.profile, config.LXD_PROFILE)
 
         LOG.debug(
             "Configured LXD substrate (profile %s, image %s)", self.profile, self.image
         )
 
-    def new_instance(self) -> Instance:
+    def new_instance(self, dualstack: bool = False) -> Instance:
         instance_id = f"k8s-integration-{os.urandom(3).hex()}-{self.next_id()}"
 
         LOG.debug("Creating instance %s with image %s", instance_id, self.image)
+        launch_lxd_command = [
+            "lxc",
+            "launch",
+            self.image,
+            instance_id,
+            "-p",
+            "default",
+            "-p",
+            self.profile,
+        ]
+
+        if dualstack:
+            # Setup dualstack profile once
+            if self.dualstack_profile is None:
+                # Create the dualstack network (fail if it already exists)
+                run(
+                    [
+                        "lxc",
+                        "network",
+                        "create",
+                        config.LXD_DUALSTACK_NETWORK,
+                        "ipv4.address=auto",
+                        "ipv6.address=auto",
+                        "ipv4.nat=true",
+                        "ipv6.nat=true",
+                    ],
+                )
+                self.dualstack_profile = config.LXD_DUALSTACK_PROFILE_NAME
+                configure_lxd_profile(
+                    self.dualstack_profile,
+                    config.LXD_DUALSTACK_PROFILE,
+                    template_overwrites={
+                        "LXD_DUALSTACK_NETWORK": config.LXD_DUALSTACK_NETWORK
+                    },
+                )
```

Comment:
i'd change this to `configure_lxd_profile(self.dualstack_profile, config.LXD_DUALSTACK_PROFILE.replace('old', 'new'))`

then, templating responsibility is on the caller, and `configure_lxd_profile only does what it says, applies it on the host